lib/scholar/interpolation/cubic_spline.ex

```diff
@@ -0,0 +1,147 @@
+      if n == 3 do
+        a =
+          Nx.concatenate(
+            [
+              Nx.tensor([[1, 1, 0]]),
+              Nx.stack([dx[1], 2 * (dx[0] + dx[1]), dx[0]]) |> Nx.new_axis(0),
+              Nx.tensor([[0, 1, 1]])
+            ],
+            axis: 0
+          )
+
+        b = Nx.stack([2 * slope[0], 3 * (dx[0] * slope[1] + dx[1] * slope[0]), 2 * slope[1]])
+
+        Nx.LinAlg.solve(a, b)
+      else
+        # n > 3 and bc = not_a_knot
+
+        d_0 = Nx.new_axis(x[2] - x[0], 0)
+        d_n = Nx.new_axis(x[-1] - x[-3], 0)
+
+        up_diag =
+          Nx.concatenate([
+            d_0,
+            dx[0..-2//1]
+          ])
+
+        low_diag =
+          Nx.concatenate([
+            dx[1..-1//1],
+            d_n
+          ])
+
+        main_diag =
+          Nx.concatenate([
+            Nx.new_axis(dx[1], 0),
+            2 * (dx[0..-2//1] + dx[1..-1//1]),
+            Nx.new_axis(dx[-2], 0)
+          ])
+
+        a =
+          Nx.broadcast(0, {n, n})
+          |> Nx.put_diagonal(up_diag, offset: 1)
+          |> Nx.put_diagonal(main_diag, offset: 0)
+          |> Nx.put_diagonal(low_diag, offset: -1)
+
+        bc_0 = ((dx[0] + 2 * d_0) * dx[1] * slope[0] + dx[0] ** 2 * slope[1]) / d_0
+
+        bc_n = (dx[-1] ** 2 * slope[-2] + (2 * d_n + dx[-1]) * dx[-2] * slope[-1]) / d_n
+
+        b =
+          Nx.concatenate([
+            Nx.new_axis(bc_0, 0),
+            3 * (dx[1..-1//1] * slope[0..-2//1] + dx[0..-2//1] * slope[1..-1//1]),
+            Nx.new_axis(bc_n, 0)
+          ])
+
+        Nx.LinAlg.solve(a, b)
+      end
+
+    slope = (y[1..-1//1] - y[0..-2//1]) / dx
+    t = (s[0..-2//1] + s[1..-1//1] - 2 * slope) / dx
+
+    c_3 = t / dx
+    c_2 = (slope - s[0..-2//1]) / dx - t
+    c_1 = s[0..-2//1]
+    c_0 = y[0..-2//1]
+
+    c = Nx.stack([c_3, c_2, c_1, c_0], axis: 0)
+
+    %__MODULE__{coefficients: c, x: x}
+  end
+
+  @doc """
+  Returns the value fit by `train/2` corresponding to the `target_x` input
+
+  ### Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+  """
+  deftransform predict(
+                 %__MODULE__{} = model,
+                 target_x,
+                 opts \\ []
+               ) do
+    predict_n(model, target_x, NimbleOptions.validate!(opts, @opts_schema))
+  end
+
+  defnp predict_n(%__MODULE__{x: x, coefficients: coefficients}, target_x, opts \\ []) do
+    nan_selector =
+      Nx.logical_and(opts[:extrapolate] == false, target_x < x[0] or target_x > x[-1])
+
+    nan = Nx.tensor(:nan, type: :f64)
+    target_x = Nx.as_type(target_x, :f64)
+
+    idx =
+      Nx.sum(x < Nx.new_axis(target_x, 1), axes: [1])
+      |> Nx.min(Nx.size(x) - 2)
+
+    c_poly = Nx.take_along_axis(coefficients, Nx.tile(idx, [4, 1]), axis: 1)
+
+    x_poly = target_x - Nx.take_along_axis(x, idx)
+
+    result =
+      x_poly
+      |> Nx.power(Nx.tensor([[3], [2], [1], [0]], type: :f64))
+      |> Nx.multiply(c_poly)
+      |> Nx.sum(axes: [0])
+
+    Nx.select(nan_selector, nan, result)
```

Comment:
We should do

```elixir
if opts[:extrapolate] do
  result
else
  nan_selector = ...
  Nx.select(...)
end
```

This way when `extrapolate: true` we will optimize-out that branch entirely (especialy that it's the default :)).